### PR TITLE
Fix tensor mechanics action material output

### PR DESCRIPTION
**Pull request by dschwen**
Refs #17606

### CODE REVIEW

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C [hunk 1]
```diff
@@ -478,17 +480,19 @@ TensorMechanicsAction::actOutputGeneration()
     unsigned int index = 0;
```
Comment (vanwdani) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:480]: You can remove this line and 496

**Comment by dschwen** [comment]
Ahhrgh, what's going on with these tests?! The exodiffing porous flow test doesn't even use tensor mechanics!

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C (hunk 1 above)
Comment (dschwen) [reply modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:480]: No, not really, its used ...

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C [hunk 2]
```diff
@@ -478,17 +480,19 @@ TensorMechanicsAction::actOutputGeneration()
...
     for (auto out : _generate_output)
     {
-      std::string type;
-      if (_material_output_order[index] == "CONSTANT")
-        type = "MooseVariableConstMonomial";
-      else
-        type = "MooseVariable";
+      const auto & order = _material_output_order[index];
+      const auto & family = _material_output_family[index];
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:484]: ... here

Comment (vanwdani) [reply modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:484]: true

**Review by vanwdani — DISMISSED** [review]

**Review by jiangwen84 — CHANGES_REQUESTED** [review]
Looks good. 

Two comments:

1. for output purpose, should we explicitly set 'execute_on = TIMESETP_END'?
2. The L2 projection approach would make a large increase of global system size when there are a lot of outputs. For some solvers, it might slow down the computation a bit. A local projection approach might be desired, which might be something to be addressed in the future.

**Comment by dschwen** [comment]
> 1. for output purpose, should we explicitly set 'execute_on = TIMESETP_END'?

See here: https://github.com/idaholab/moose/blob/51d56fe51e93bcbb34a1d0a4599d46a2ef8a55cf/modules/tensor_mechanics/src/actions/TensorMechanicsAction.C#L520

> 2. The L2 projection approach would make a large increase of global system size when there are a lot of outputs.

Yeah, agree. A custom preconditioner or split preconditioner could help as well.

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C [hunk 3]
```diff
@@ -478,17 +480,19 @@ TensorMechanicsAction::actOutputGeneration()
...
+
       // Create output helper aux variables
       auto params = _factory.getValidParams(type);
-      params.set<MooseEnum>("order") = _material_output_order[index];
-      params.set<MooseEnum>("family") = _material_output_family[index];
+      params.set<MooseEnum>("order") = order;
+      params.set<MooseEnum>("family") = family;
 
-      if (_material_output_family[index] == "MONOMIAL")
+      if (family == "MONOMIAL")
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:495]: Should we consider only `MONOMIAL` for `AuxVariable` and `LAGRANGE` for `NonlinearVariable` as output options ? I am not sure if the `L2 project` would work for any other types of interpolation function. Maybe some but not all. Throw a `mooseError`?

Comment (dschwen) [reply modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:495]: Yeah, prudent to limit this until we verify.

Comment (dschwen) [reply modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:495]: Thinking a bit more about this, why wouldn't arbitrary shape functions work when outputting via a nonlinear variable?

Comment (jiangwen84) [reply modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:495]: I might be wrong, but if you take a look at all family type 

` MooseEnum family("LAGRANGE MONOMIAL HERMITE SCALAR HIERARCHIC CLOUGH XYZ SZABAB BERNSTEIN "
                   "L2_LAGRANGE L2_HIERARCHIC NEDELEC_ONE LAGRANGE_VEC MONOMIAL_VEC",
                   "LAGRANGE");`

`HERMITE` I believe only works for structure mesh (not very sure though) and undisplaced mesh. `SCALAR` won't work neither. I am not familiar with other types, such as `CLOUGH`, `BERNSTEIN`... , so I do not know if those work not or.

**Comment by dschwen** [comment]
Yeah, but for escalate and vector elements etc. you will already get an
error from moose...

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C (hunk 3 above)
Comment (dschwen) [reply modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:495]: HERMITE actually gives `PC failed due to FACTOR_NUMERIC_ZEROPIVOT`. Ok, I'm restricting it to `MONOMIAL` and `LAGRANGE`

**Review by jiangwen84 — APPROVED** [review]

**Pull request merged**